File: sdk/c/examples/node.cpp
// Generic Matrix test node for C++ (ML1 contract: docs/ML1-NODE.md).
//
// Usage: mx-node-cpp --matrix-sock <sock> --id <logical>
//        [--event-log <path>] [--stream-log <path>] [--stream-slow-ms <n>]
#include "matrix.hpp"

#include <chrono>
#include <cstdint>
#include <cstdio>
#include <cstring>
#include <fstream>
#include <string>
#include <thread>

namespace {

std::string g_id = "dep-node";
std::string g_event_log;
std::string g_stream_log;
long g_stream_slow_ms = 0;

void append_line(const std::string &path, const std::string &line) {
    if (path.empty())
        return;
    std::ofstream f(path, std::ios::app);
    if (f)
        f << line << "\n";
}

// Top-level member lookup over the raw input. The node only needs a few
// shapes, but nested objects (chain `input`, stream specs) may repeat a
// top-level key (e.g. spec `sleep_ms` shadowing the sleep branch), so
// first-occurrence search is wrong: only depth-1 members match. Strings
// (with backslash escapes) are skipped while scanning.
std::size_t find_top(const std::string &in, const std::string &key) {
    int depth = 0;
    bool in_str = false;
    for (std::size_t i = 0; i < in.size(); ++i) {
        char c = in[i];
        if (in_str) {
            if (c == '\\')
                ++i;
            else if (c == '"')
                in_str = false;
            continue;
        }
        if (c == '"') {
            if (depth == 1 && in.compare(i + 1, key.size(), key) == 0 &&
                i + 2 + key.size() <= in.size() && in[i + 1 + key.size()] == '"') {
                std::size_t j = i + 2 + key.size();
                while (j < in.size() && (in[j] == ' ' || in[j] == '\t' || in[j] == '\n' || in[j] == '\r'))
                    ++j;
                if (j < in.size() && in[j] == ':')
                    return j + 1;
            }
            in_str = true;
            continue;
        }
        if (c == '{' || c == '[')
            ++depth;
        else if (c == '}' || c == ']')
            --depth;
    }
    return std::string::npos;
}
static std::size_t skip_ws(const std::string &in, std::size_t i) {
    while (i < in.size() && (in[i] == ' ' || in[i] == '\t' || in[i] == '\n' || in[i] == '\r'))
        ++i;
    return i;
}
bool has_true(const std::string &in, const std::string &key) {
    auto p = find_top(in, key);
    if (p == std::string::npos)
        return false;
    return in.compare(skip_ws(in, p), 4, "true") == 0;
}
std::string get_str(const std::string &in, const std::string &key) {
    auto p = find_top(in, key);
    if (p == std::string::npos)
        return "";
    p = skip_ws(in, p);
    if (p >= in.size() || in[p] != '"')
        return "";
    std::string o;
    for (std::size_t i = p + 1; i < in.size(); ++i) {
        if (in[i] == '\\' && i + 1 < in.size()) {
            o += in[i + 1];
            ++i;
        } else if (in[i] == '"') {
            return o;
        } else {
            o += in[i];
        }
    }
    return "";
}
long long get_int(const std::string &in, const std::string &key, long long dflt) {
    auto p = find_top(in, key);
    if (p == std::string::npos)
        return dflt;
    try {
        return std::stoll(in.substr(skip_ws(in, p)));
    } catch (...) {
        return dflt;
    }
}
// Numeric-only u64 (Rust dep_node reference: only a non-negative
// integer JSON number takes the release branch; strings, floats,
// bools and bad shapes fall through to echo).
bool get_release(const std::string &in, std::uint64_t &out) {
    auto p = find_top(in, "release");
    if (p == std::string::npos)
        return false;
    p = skip_ws(in, p);
    if (p >= in.size() || in[p] < '0' || in[p] > '9')
        return false;
    std::size_t q = p;
    while (q < in.size() && in[q] >= '0' && in[q] <= '9')
        ++q;
    if (q < in.size() && (in[q] == '.' || in[q] == 'e' || in[q] == 'E'))
        return false;
    try {
        std::size_t pos = 0;
        unsigned long long v = std::stoull(in.substr(p, q - p), &pos);
        if (pos != q - p)
            return false;
        out = static_cast<std::uint64_t>(v);
        return true;
    } catch (...) {
        return false;
    }
}
std::string sub_object(const std::string &in, const std::string &key) {
    auto p = find_top(in, key);
    if (p == std::string::npos)
        return "{}";
    p = skip_ws(in, p);
    if (p >= in.size() || in[p] != '{')
        return "{}";
    int depth = 0;
    bool in_str = false;
    for (std::size_t i = p; i < in.size(); ++i) {
        char c = in[i];
        if (in_str) {
            if (c == '\\')
                ++i;
            else if (c == '"')
                in_str = false;
            continue;
        }
        if (c == '"') {
            in_str = true;
        } else if (c == '{') {
            ++depth;
        } else if (c == '}') {
            if (--depth == 0)
                return in.substr(p, i - p + 1);
        }
    }
    return "{}";
}
std::string quote(const std::string &s) {
    std::string o = "\"";
    for (char ch : s) {
        if (ch == '"' || ch == '\\')
            o += '\\';
        o += ch;
    }
    return o + "\"";
}

class Node : public mx::Handler {
public:
    std::string on_call(mx::CallCtx &ctx, const std::string &,
                        const std::string &, const std::string &input) override {
        long long sleep_ms = get_int(input, "sleep_ms", 0);
        for (long long s = 0; s < sleep_ms; s += 5) {
            if (ctx.cancelled())
                throw mx::BusinessError("cancelled", "aborted");
            std::this_thread::sleep_for(std::chrono::milliseconds(5));
        }
        std::string fail = get_str(input, "fail");
        if (!fail.empty())
            throw mx::BusinessError(fail, "remote " + fail);
        // amplify / chain / acquire / release / streams mirror node.c.
        if (find_top(input, "amplify") != std::string::npos) {
            long long n = get_int(input, "amplify", 0);
            n = std::max<long long>(0, std::min<long long>(n, 1 << 20));
            return "{\"blob\":\"" + std::string(static_cast<std::size_t>(n), 'x') +
                   "\",\"via\":" + quote(g_id) + "}";
        }
        if (has_true(input, "chain")) {
            auto deps = ctx.dependencies();
            if (deps.empty())
                throw mx::BusinessError("dependency-unavailable", "no binding");
            long long timeout = std::max<long long>(1, get_int(input, "timeout_ms", 5000));
            std::string out = ctx.invoke_dependency(deps[0].id, sub_object(input, "input"),
                                                    static_cast<unsigned long long>(timeout));
            return "{\"chained\":" + out + ",\"via\":" + quote(g_id) + "}";
        }
        if (find_top(input, "acquire") != std::string::npos) {
            std::string acq = sub_object(input, "acquire");
            long long ms = get_int(acq, "interval_ms", -1);
            std::uint64_t h = ctx.acquire_resource(get_str(acq, "kind"), get_str(acq, "label"),
                                                   ms >= 0, static_cast<unsigned long long>(ms));
            return "{\"acquired\":{\"handle\":\"" + std::to_string(h) +
                   "\"},\"via\":" + quote(g_id) + "}";
        }
        if (find_top(input, "release") != std::string::npos) {
            std::uint64_t h = 0;
            if (get_release(input, h)) {
                ctx.release_resource(h);
                return "{\"released\":\"" + std::to_string(h) + "\",\"via\":" + quote(g_id) + "}";
            }
        }
        if (find_top(input, "stream_send") != std::string::npos) {
            std::string spec = sub_object(input, "stream_send");
            std::string sid = get_str(spec, "stream_id");
            if (sid.empty())
                sid = "s-test";
            long long chunks = std::min<long long>(std::max<long long>(0, get_int(spec, "chunks", 0)), 256);
            long long nbytes = std::min<long long>(std::max<long long>(0, get_int(spec, "chunk_bytes", 0)), 4096);
            long long slp = std::max<long long>(0, get_int(spec, "sleep_ms", 0));
            std::string payload(static_cast<std::size_t>(nbytes), 'x');
            long long sent = 0;
            for (long long seq = 0; seq < chunks; ++seq) {
                if (ctx.cancelled())
                    throw mx::BusinessError("cancelled", "aborted");
                try {
                    ctx.send_stream(sid, static_cast<std::uint64_t>(seq), payload);
                } catch (const mx::Error &e) {
                    throw mx::BusinessError("stream-refused", e.what());
                }
                ++sent;
                if (slp > 0) {
                    long long s = std::min(slp, 50LL);
                    for (long long w = 0; w < s; w += 5) {
                        if (ctx.cancelled())
                            throw mx::BusinessError("cancelled", "aborted");
                        std::this_thread::sleep_for(std::chrono::milliseconds(5));
                    }
                }
            }
            return "{\"stream_sent\":" + std::to_string(sent) + ",\"via\":" + quote(g_id) + "}";
        }
        if (find_top(input, "chain_with_streams") != std::string::npos) {
            // Concurrent chain + streams (M7 bidi legs): streams while
            // the child leg is in flight on this same session.
            std::string spec = sub_object(input, "chain_with_streams");
            std::string sid = get_str(spec, "stream_id");
            if (sid.empty())
                sid = "s-bidi";
            long long chunks = std::min<long long>(std::max<long long>(0, get_int(spec, "chunks", 0)), 32);
            long long nbytes = std::min<long long>(std::max<long long>(0, get_int(spec, "chunk_bytes", 0)), 1024);
            long long interval = std::min<long long>(std::max<long long>(0, get_int(spec, "interval_ms", 20)), 50);
            long long prime = std::min<long long>(std::max<long long>(0, get_int(spec, "prime_ms", 50)), 1000);
            std::string payload(static_cast<std::size_t>(nbytes), 'x');
            long long sent = 0;
            std::thread streamer([&] {
                if (prime > 0)
                    std::this_thread::sleep_for(std::chrono::milliseconds(prime));
                for (long long seq = 0; seq < chunks; ++seq) {
                    try {
                        ctx.send_stream(sid, static_cast<std::uint64_t>(seq), payload);
                        ++sent;
                    } catch (...) {
                        break;
                    }
                    if (interval > 0)
                        std::this_thread::sleep_for(std::chrono::milliseconds(interval));
                }
            });
            auto deps = ctx.dependencies();
            if (deps.empty()) {
                streamer.join();
                throw mx::BusinessError("dependency-unavailable", "no binding");
            }
            long long timeout = std::max<long long>(1, get_int(spec, "timeout_ms", 8000));
            std::string chainedOut;
            try {
                chainedOut = ctx.invoke_dependency(deps[0].id, sub_object(spec, "input"),
                                                   static_cast<unsigned long long>(timeout));
            } catch (const mx::Error &e) {
                streamer.join();
                throw mx::BusinessError(e.code, e.what());
            }
            streamer.join();
            return "{\"chained\":" + chainedOut + ",\"via\":" + quote(g_id) +
                   ",\"stream_sent\":" + std::to_string(sent) + "}";
        }
        return "{\"echo\":" + input + ",\"via\":" + quote(g_id) + "}";
    }

    void on_event(const std::string &topic, const std::string &payload) override {
        append_line(g_event_log, topic + "\t" + payload);
    }
    void on_stream(const std::string &id, std::uint64_t seq,
                   const std::string &payload) override {
        if (g_stream_slow_ms > 0)
            std::this_thread::sleep_for(std::chrono::milliseconds(g_stream_slow_ms));
        append_line(g_stream_log, id + "\t" + std::to_string(seq) + "\t" +
                                      std::to_string(payload.size()));
    }
};

} // namespace

int main(int argc, char **argv) {
    std::string sock;
    for (int i = 1; i < argc; ++i) {
        std::string a = argv[i];
        auto need = [&](const char *k, std::string &out) {
            if (a == k && i + 1 < argc)
                out = argv[++i];
        };
        need("--matrix-sock", sock);
        need("--id", g_id);
        need("--event-log", g_event_log);
        need("--stream-log", g_stream_log);
        if (a == "--stream-slow-ms" && i + 1 < argc)
            g_stream_slow_ms = std::stoll(argv[++i]);
    }
    if (sock.empty()) {
        std::fprintf(stderr, "usage: mx-node-cpp --matrix-sock <sock> [--id <logical>] ...\n");
        return 2;
    }
    try {
        Node handler;
        mx::Component comp = mx::Component::connect(sock, g_id, handler);
        std::string reason = comp.serve();
        comp.close();
        if (reason != "dispose" && reason != "eof") {
            std::fprintf(stderr, "serve: %s\n", reason.c_str());
            return 1;
        }
        return 0;
    } catch (const std::exception &e) {
        std::fprintf(stderr, "fatal: %s\n", e.what());
        return 1;
    }
}

File: sdk/js/examples/node.js
"use strict";
/* Generic Matrix test node (ML1 contract: docs/ML1-NODE.md).
 * Usage: node node.js --matrix-sock <sock> --id <logical>
 *        [--event-log <path>] [--stream-log <path>] [--stream-slow-ms <n>]
 */
const fs = require("node:fs");
const { Component, Handler } = require("../lib/component");

function appendLine(p, line) {
  try {
    fs.appendFileSync(p, line + "\n");
  } catch { /* log loss never fails the call */ }
}

function sleepMs(ms, signal) {
  return new Promise((resolve, reject) => {
    if (signal.aborted) return reject(Object.assign(new Error("aborted"), { code: "cancelled" }));
    const t = setTimeout(() => {
      signal.removeEventListener("abort", onAbort);
      resolve(false);
    }, ms);
    const onAbort = () => {
      clearTimeout(t);
      reject(Object.assign(new Error("aborted"), { code: "cancelled" }));
    };
    signal.addEventListener("abort", onAbort, { once: true });
  });
}

async function abortableSleep(totalMs, signal) {
  let slept = 0;
  while (slept < totalMs) {
    const step = Math.min(5, totalMs - slept);
    await sleepMs(step, signal);
    slept += step;
  }
}

class Node extends Handler {
  constructor(id, eventLog, streamLog, streamSlowMs) {
    super();
    this.id = id;
    this.eventLog = eventLog;
    this.streamLog = streamLog;
    this.streamSlowMs = streamSlowMs;
  }
  onEvent(topic, payload) {
    if (this.eventLog) appendLine(this.eventLog, `${topic}\t${JSON.stringify(payload)}`);
  }
  async onStream(streamId, seq, payload) {
    if (this.streamSlowMs > 0) await new Promise((r) => setTimeout(r, this.streamSlowMs));
    if (this.streamLog) appendLine(this.streamLog, `${streamId}\t${seq}\t${payload.length}`);
  }
  async onCall(ctx, ticket, cap, input, signal) {
    input = input && typeof input === "object" ? input : {};
    const sleepN = Number(input.sleep_ms || 0);
    if (sleepN > 0) {
      try {
        await abortableSleep(sleepN, signal);
      } catch (e) {
        throw Object.assign(new Error("aborted"), { code: "cancelled" });
      }
    }
    if (typeof input.fail === "string" && input.fail)
      throw Object.assign(new Error(`remote ${input.fail}`), { code: input.fail });
    if (Number.isInteger(input.amplify)) {
      const n = Math.min(Math.max(input.amplify, 0), 1 << 20);
      return { blob: "x".repeat(n), via: this.id };
    }
    if (input.chain) {
      const bindings = ctx.dependencies();
      if (!bindings.length) throw Object.assign(new Error("no binding"), { code: "dependency-unavailable" });
      const inner = input.input && typeof input.input === "object" ? input.input : {};
      const rawTimeout = Number(input.timeout_ms ?? 5000);
      const timeoutMs = Math.max(Number.isFinite(rawTimeout) ? rawTimeout : 5000, 1);
      const out = await ctx.invokeDependency(bindings[0].id, inner, timeoutMs / 1000);
      return { chained: out, via: this.id };
    }
    if (input.acquire && typeof input.acquire === "object") {
      const acq = input.acquire;
      const ms = acq.interval_ms === undefined || acq.interval_ms === null ? undefined : Number(acq.interval_ms);
      const h = await ctx.acquireResource(String(acq.kind || ""), String(acq.label || ""), ms);
      return { acquired: { handle: h.toString() }, via: this.id };
    }
    const rel = input.release;
    if ((typeof rel === "number" && Number.isInteger(rel) && rel >= 0) ||
        (typeof rel === "bigint" && rel >= 0n)) {
      await ctx.releaseResource(rel);
      return { released: String(rel), via: this.id };
    }
    if (input.stream_send && typeof input.stream_send === "object") {
      const spec = input.stream_send;
      const streamId = String(spec.stream_id || "s-test");
      const chunks = Math.min(Math.max(Number(spec.chunks || 0), 0), 256);
      const nbytes = Math.min(Math.max(Number(spec.chunk_bytes || 0), 0), 4096);
      const slp = Math.max(Number(spec.sleep_ms || 0), 0);
      const payload = "x".repeat(nbytes);
      let sent = 0;
      for (let seq = 0; seq < chunks; seq++) {
        if (signal.aborted) throw Object.assign(new Error("aborted"), { code: "cancelled" });
        try {
          await ctx.sendStream(streamId, seq, payload);
        } catch (e) {
          throw Object.assign(new Error(String(e.detail || e.message)), { code: "stream-refused" });
        }
        sent++;
        if (slp > 0) await abortableSleep(Math.min(slp, 50), signal).catch(() => {
          throw Object.assign(new Error("aborted"), { code: "cancelled" });
        });
      }
      return { stream_sent: sent, via: this.id };
    }
    if (input.chain_with_streams && typeof input.chain_with_streams === "object") {
      // Concurrent chain + streams (M7 bidi legs): streams while the
      // child leg is in flight on this same session.
      const spec = input.chain_with_streams;
      const streamId = String(spec.stream_id || "s-bidi");
      const chunks = Math.min(Math.max(Number(spec.chunks || 0), 0), 32);
      const nbytes = Math.min(Math.max(Number(spec.chunk_bytes || 0), 0), 1024);
      const interval = Math.min(Math.max(Number(spec.interval_ms ?? 20), 0), 50);
      const prime = Math.min(Math.max(Number(spec.prime_ms ?? 50), 0), 1000);
      const payload = "x".repeat(nbytes);
      let sent = 0;
      const streamer = (async () => {
        if (prime > 0) await new Promise((r) => setTimeout(r, prime));
        for (let seq = 0; seq < chunks; seq++) {
          try {
            await ctx.sendStream(streamId, seq, payload);
            sent++;
          } catch {
            break;
          }
          if (interval > 0) await new Promise((r) => setTimeout(r, interval));
        }
      })();
      const bindings = ctx.dependencies();
      if (!bindings.length) {
        await streamer;
        throw Object.assign(new Error("no binding"), { code: "dependency-unavailable" });
      }
      const inner = spec.input && typeof spec.input === "object" ? spec.input : {};
      const rawTimeout = Number(spec.timeout_ms ?? 8000);
      const timeoutMs = Math.max(Number.isFinite(rawTimeout) ? rawTimeout : 8000, 1);
      let chainedOut;
      try {
        chainedOut = await ctx.invokeDependency(bindings[0].id, inner, timeoutMs / 1000);
      } catch (e) {
        await streamer;
        throw Object.assign(new Error(String(e.detail || e.message)), { code: e.code || "internal" });
      }
      await streamer;
      return { chained: chainedOut, via: this.id, stream_sent: sent };
    }
    return { echo: input, via: this.id };
  }
}

async function main() {
  const args = process.argv.slice(2);
  const get = (k) => {
    const i = args.indexOf(k);
    return i >= 0 ? args[i + 1] : undefined;
  };
  const sock = get("--matrix-sock");
  const id = get("--id") || "dep-node";
  if (!sock) {
    process.stderr.write("usage: node.js --matrix-sock <sock> [--id <logical>] ...\n");
    process.exit(2);
  }
  let comp;
  try {
    comp = await Component.connect(sock, id);
  } catch (e) {
    process.stderr.write(`connect: ${e.message}\n`);
    process.exit(2);
  }
  const reason = await comp.serve(new Node(id, get("--event-log") || null,
    get("--stream-log") || null, Number(get("--stream-slow-ms") || 0)));
  if (reason !== "dispose" && reason !== "eof") {
    process.stderr.write(`serve: ${reason}\n`);
    process.exit(1);
  }
}

if (require.main === module) main().catch((e) => {
  process.stderr.write(`fatal: ${e.stack || e}\n`);
  process.exit(1);
});
module.exports = { Node };
